management_api.go

```diff
@@ -122,6 +122,22 @@ func (e *Enforcer) AddPolicy(params ...interface{}) (bool, error) {
 	return e.AddNamedPolicy("p", params...)
 }
 
+// AddPolicies adds authorization rules to the current policy.
+// If the rule already exists, the function returns false for the corresponding rule and the rule will not be added.
+// Otherwise the function returns true for the corresponding rule by adding the new rule.
+func (e *Enforcer) AddPolicies(rules ...interface{}) [][] interface{} {
```

Comment:
I think using `[][]interface{}` is more better.
WDYT?

Reply:
Actually, it would be better if we use like this, don't you think as the users can always use spread operator if they have rules in ```[][] interface{}```, like
```
rules = [][]interface{} {
                   []interface{} {"jack", "data4", "read"},
                   []interface{} {"katy", "data4", "write"},
                   []interface{} {"leyo", "data4", "read"},
                   []interface{} {"ham", "data4", "write"}
                }
```
And then can be used like,
```
e.AddPolicies(rules...)
e.RemovePolicies(rules...)
```

Reply:
What do you think @nodece should I leave this as it is, or should I make changes as you said?

Reply:
What's the difference between []interface{} and [][]interface{}?

Reply:
Hi @hsluoyz, the ```[] interface{} ``` can consists of ```array``` or ```primitive datatypes``` like ```string```, ```int``` or ```float```, but the ```[][]interface{} array``` must only consists of any ```nested array```, it cannot contain a ```primitive data-type```.
Right.

Reply:
I pretty, like:
```
func (e *Enforcer) AddPolicies(params [][]interface{}) (bool, error)
```
```
rules := [][]interface{}{
	{"user","/","get"},
	{"user", "/", "post"},
	{"user","/","put"},
}

e.AddPolicies(rules)
```

Reply:
Then I agree with @nodece

Reply:
I forgot that `interface{}` can contain `arrays`.

Reply:
I actually tried maintaining the structure of the casbin, as every function used variadic arguments, so I thought it would be better if I use such arguments instead of passing a 2-D array. But its okay @nodece I will make the changes if @hsluoyz is also okay with this.
Thanks @nodece and @hsluoyz.

Reply:
@hsluoyz Why don't we use `[][]string`?

Reply:
Each policy  is `[]string`.

Reply:
```
rule1 := []interface{} {"jack", "data4", "read"}
rule2 := []interface{} {"katy", "data4", "write"}
rule3 := []interface{} {"leyo", "data4", "read"}
rule4 := []interface{} {"ham", "data4", "write"}

e.AddPolicies(rule1, rule2, rule3, rule4)

e.RemovePolicies(rule1, rule2, rule3, rule4)
```
What do you think, in this case, variadic arguments won't be better?

Reply:
> Each policy is `[]string`.

Oh, yeah thats better, I feel.

Reply:
Agreed @nodece this is better signature.
```func AddPolicies(rules ...[]string) [][]interface```

Reply:
Then, I think this must also be changed in the management_api.go
```func (e *Enforcer) AddPolicy(params ...interface{}) (bool, error) ``` to ```func (e *Enforcer) AddPolicy(params ...string) (bool, error) ```

Reply:
> ```
> rule1 := []interface{} {"jack", "data4", "read"}
> rule2 := []interface{} {"katy", "data4", "write"}
> rule3 := []interface{} {"leyo", "data4", "read"}
> rule4 := []interface{} {"ham", "data4", "write"}
> 
> e.AddPolicies(rule1, rule2, rule3, rule4)
> 
> e.RemovePolicies(rule1, rule2, rule3, rule4)
> ```
> 
> What do you think, in this case, variadic arguments won't be better?

It's not friendly for users. users pretty to pass 2D array.

Reply:
If a user wants to pass 2D array, then it can be used like this ```e.AddPolicies(rules...)```

Reply:
Its best of both the worlds.

Reply:
@nodece should I make changes from interface{} to string ?

Reply:
If I'm a user, I want to use `e.AddPolicies(rules)` instead of `e.AddPolicies(rules...)`

Reply:
> @nodece should I make changes from interface{} to string ?

I want to hear what @hsluoyz.

Reply:
> @nodece should I make changes from interface{} to string ?

I want to listen from @hsluoyz .

Reply:
No one will call `e.AddPolicies(rule1, rule2, rule3, rule4)`.

`e.AddPolicies(rules)` is better than `e.AddPolicies(rules...)`

Reply:
The `[][]string` as parameter is better. @hsluoyz WDYT?